File: Best Time to Buy and Sell Stock III.cpp
class Solution {
public:
    int maxProfit(vector<int> &prices) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        vector<int>historyProfits(prices.size(), 0);
        int buy = 0;
        int max = 0;
        int profit = 0;
        
        int N = prices.size();
        for(int i = 0; i < N; i++)
        {
            buy = prices[i]<prices[buy]? i:buy;
            profit = prices[i] - prices[buy];
            max = profit>max? profit:max;
            historyProfits[i] = max;
        }
        
        int sell = N-1;
        max = 0;
        profit = 0;
        for(int i = N-1; i>-1; i--)
        {
            sell = prices[i]>prices[sell]? i:sell;
            profit = prices[sell] - prices[i]+historyProfits[i];
            max = profit>max? profit:max;
        }
        
        return max;
    }
};

File: Best Time to Buy and Sell Stock III.java
public class Solution {
    public int maxProfit(int[] prices) {
        // Start typing your Java solution below
        // DO NOT write main() function
        int []buy = new int[prices.length];
        int maxVal = Integer.MIN_VALUE;
        int maxprofit = 0;
        for(int i = prices.length-1; i >-1 ; i--)
        {
            if(prices[i] > maxVal)
            {
                maxVal = prices[i];    
            } 
            int val = maxVal - prices[i];
            maxprofit = Math.max(val, maxprofit);
            buy[i] = maxprofit; 
        } 
        
        maxprofit = 0; 
        int minVal = Integer.MAX_VALUE;
        for(int i = 0; i < prices.length; i++)
        {
            if(prices[i] < minVal)
            {
                minVal = prices[i];
            } 
            int val = prices[i] - minVal;
            maxprofit = Math.max(maxprofit, val + buy[i]);
        }
        
        return maxprofit;
    }
}